lc-hax/Scripts/Modules/Possession/PossessionMod.cs

```diff
@@ -190,15 +233,110 @@ internal sealed class PossessionMod : MonoBehaviour {
+&& this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value) && value.CanUseEntranceDoors(enemyAI);
+    }
+
+
+    float InteractRange(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? 0
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value)
+            ? value.InteractRange(enemyAI).GetValueOrDefault(1.0f)
+            : 1.0f;
+    }
+
+
+
+    void InteractWithAmbient() {
+        if (this.doorCooldownRemaining > 0) this.doorCooldownRemaining -= Time.deltaTime;
+        if (this.teleportCooldownRemaining > 0) this.teleportCooldownRemaining -= Time.deltaTime;
+        if (this.Possession.Enemy is not EnemyAI enemy) return;
+
+
+        if (Physics.Raycast(enemy.transform.position, enemy.transform.forward, out RaycastHit hit, this.InteractRange(enemy))) {
+            if (hit.collider.gameObject.TryGetComponent(out DoorLock doorLock) && this.doorCooldownRemaining <= 0) {
+                this.OpenDoorAsEnemy(doorLock);
+                this.doorCooldownRemaining = DoorInteractionCooldown;
+                return;
+            }
+
+            if (this.CanUseEntranceDoors(enemy)) {
+                if (hit.collider.gameObject.TryGetComponent(out EntranceTeleport entrance) &&
+                    this.teleportCooldownRemaining <= 0) {
+                    this.InteractWithTeleport(enemy, entrance);
+                    this.teleportCooldownRemaining = TeleportDoorCooldown;
+                    return;
+                }
+            }
+        }
+    }
+
+    void OpenDoorAsEnemy(DoorLock door) {
+        if (!door.Reflect().GetInternalField<bool>("isDoorOpened")) {
+            door.OpenDoorAsEnemyServerRpc();
+            if (door.gameObject.TryGetComponent(out AnimatedObjectTrigger trigger)) {
+                trigger.TriggerAnimationNonPlayer(false, true, false);
+            }
+        }
+    }
+
+
+    Transform? GetExitPointFromDoor(EntranceTeleport entrance) {
+        if (entrance == null) {
+            return null;
+        }
+        EntranceTeleport[] allTeleports = FindObjectsOfType<EntranceTeleport>();
+        for (int index = 0; index < allTeleports.Length; index++) {
+            EntranceTeleport teleport = allTeleports[index];
+            if (teleport.entranceId == entrance.entranceId &&
+                teleport.isEntranceToBuilding != entrance.isEntranceToBuilding) {
+                return teleport.entrancePoint;
+            }
+        }
+
+        return null;
+    }
+
+
+
+    public void InteractWithTeleport(EnemyAI Enemy, EntranceTeleport teleport) {
+        if (this.CharacterMovement is not CharacterMovement characterMovement) return;
+        Transform? exitPoint = this.GetExitPointFromDoor(teleport);
+        if (exitPoint == null) return;
+        Enemy.isOutside = !teleport.isEntranceToBuilding;
+        Enemy.allAINodes = Enemy.isOutside ? GameObject.FindGameObjectsWithTag("OutsideAINode") : GameObject.FindGameObjectsWithTag("AINode");
```

Comment:
Oh this is interesting. Did you find this with UnityExplorer?

Reply:
A little bit of both, i saw how lethal escape works and i saw how it sets the outside nodes